NetKet/Machine/mps_diagonal.hpp

```diff
@@ -0,0 +1,743 @@
+    std::size_t nchange = tochange.size();
+    if (nchange <= 0) {
+      return;
+    }
+    std::vector<std::size_t> sorted_ind = sort_indeces(tochange);
+    int site = tochange[sorted_ind[0]];
+
+    // Update left (site++)
+    if (site == 0) {
+      lt.V(0) = W_[0][confindex_[newconf[sorted_ind[0]]]];
+    } else {
+      lt.V(2 * site) =
+          lt.V(2 * (site - 1))
+              .cwiseProduct(
+                  W_[site % symperiod_][confindex_[newconf[sorted_ind[0]]]]);
+    }
+
+    for (std::size_t k = 1; k < nchange; k++) {
+      for (site = tochange[sorted_ind[k - 1]] + 1;
+           site < tochange[sorted_ind[k]]; site++) {
+        lt.V(2 * site) =
+            lt.V(2 * (site - 1))
+                .cwiseProduct(W_[site % symperiod_][confindex_[v(site)]]);
+      }
+      site = tochange[sorted_ind[k]];
+      lt.V(2 * site) =
+          lt.V(2 * (site - 1))
+              .cwiseProduct(
+                  W_[site % symperiod_][confindex_[newconf[sorted_ind[k]]]]);
+    }
+
+    for (site = tochange[sorted_ind[nchange - 1]] + 1; site < N_; site++) {
+      lt.V(2 * site) =
+          lt.V(2 * (site - 1))
+              .cwiseProduct(W_[site % symperiod_][confindex_[v(site)]]);
+    }
+
+    // Update right (site--)
+    site = tochange[sorted_ind[nchange - 1]];
+    if (site == N_ - 1) {
+      lt.V(1) = W_[(N_ - 1) % symperiod_]
+                  [confindex_[newconf[sorted_ind[nchange - 1]]]];
+    } else {
+      lt.V(2 * (N_ - site) - 1) =
+          W_[site % symperiod_][confindex_[newconf[sorted_ind[nchange - 1]]]]
+              .cwiseProduct(lt.V(2 * (N_ - site) - 3));
+    }
+
+    for (int k = nchange - 2; k >= 0; k--) {
+      for (site = tochange[sorted_ind[k + 1]] - 1;
+           site > tochange[sorted_ind[k]]; site--) {
+        lt.V(2 * (N_ - site) - 1) =
+            W_[site % symperiod_][confindex_[v(site)]].cwiseProduct(
+                lt.V(2 * (N_ - site) - 3));
+      }
+      site = tochange[sorted_ind[k]];
+      lt.V(2 * (N_ - site) - 1) =
+          W_[site % symperiod_][confindex_[newconf[sorted_ind[k]]]]
+              .cwiseProduct(lt.V(2 * (N_ - site) - 3));
+    }
+
+    for (site = tochange[sorted_ind[0]] - 1; site >= 0; site--) {
+      lt.V(2 * (N_ - site) - 1) =
+          W_[site % symperiod_][confindex_[v(site)]].cwiseProduct(
+              lt.V(2 * (N_ - site) - 3));
+    }
+  };
+
+  // Auxiliary function that calculates contractions from site1 to site2
+  inline VectorType mps_contraction(const Eigen::VectorXd &v, const int &site1,
+                                    const int &site2) {
+    VectorType c = VectorType::Ones(D_);
+    for (int site = site1; site < site2; site++) {
+      c = c.cwiseProduct(W_[site % symperiod_][confindex_[v(site)]]);
+    }
+    return c;
+  };
+
+  T LogVal(const Eigen::VectorXd &v) override {
+    return std::log(mps_contraction(v, 0, N_).sum());
+  };
+
+  T LogVal(const Eigen::VectorXd &v, const LookupType &lt) override {
```

Comment:
here the compiler complains about the variable ```v``` being unused. The best trick to remove the warning is to do
```
T LogVal(const Eigen::VectorXd & /* v */, const LookupType &lt) ... 
```